feat: add Ubiquity Pool UI

zugdev opened the pull request:
Resolves #1

zugdev commented:
Not sure if we should express the minting fee in percentage, UUSD or USD. In my opinion, percentage plus USD makes sense to me.

rndquu commented:
Related [comment](https://github.com/ubiquity/uusd.ubq.fi/issues/1#issuecomment-2465092704). Closing current PR (temporarily?).

zugdev commented:
## Full flow demo

https://github.com/user-attachments/assets/ac3c4347-f1eb-41da-824d-22208f7e1458

## To activate anvil

1. Check readme.md

2. Switch network

https://github.com/user-attachments/assets/b2901b13-c428-4f4b-a8f2-d5b8dbeaf204

zugdev commented:
@rndquu this works, you can try with anvil. I will be improving some UI edges over time, but that should not stop you from reviewing. I want to import the toaster as success and error.

0x4007 commented:
@rndquu  please lead the review

rndquu reviewed (changes requested):
Overall looks good. There're a couple of UI bugs:
1. Somehow 2 txs are sent in this case in the "Mint" page:
  a) User clicks "Mint" (1 mint tx is sent)
  b) User checks the "Force Collateral Only" toggle and clicks "Mint" again (in this case 2 txs are sent)
<img width="1512" alt="Screenshot 2025-02-03 at 17 22 12" src="https://github.com/user-attachments/assets/41cc606f-b2a8-450c-bd0e-047713706be5" />
2. Can't focus text input:
<img width="1512" alt="Screenshot 2025-02-03 at 17 25 38" src="https://github.com/user-attachments/assets/cf561094-b67a-40ac-9744-3e7bd4ea905a" />
3. If I switch to ethereum mainnet then spot prices are not shown + there're tons of CORS errors in the console:
<img width="1512" alt="Screenshot 2025-02-03 at 17 31 27" src="https://github.com/user-attachments/assets/6bb61f3f-0108-48bf-b24b-2736996cccd6" />
4. I don't fully understand why on sending txs metamask doesn't ask to sign anything? It seems there's some kind of anvil auto impersonation enabled.

zugdev commented:
Thank you very much for the very detailed QA.

1. Fixed

2. Fixed

3. This happens when we get rate limited by RPC provider, here is a screenshot where I am connected to Ethereum and prices are correctly fetched.

<img width="1106" alt="Screenshot 2025-02-04 at 18 59 12" src="https://github.com/user-attachments/assets/41eba2a4-132a-4a40-978d-3b0abc09c857" />

4. Yes, that's weird im still investigating.

zugdev commented:
https://f72b346f.uusd-ubq-fi.pages.dev/

rndquu reviewed (changes requested):
In the "Redeem" page on clicking the "Approve UUSD" button this error occurs:
<img width="1512" alt="Screenshot 2025-02-05 at 11 09 53" src="https://github.com/user-attachments/assets/a0d32fb9-0e63-4b6b-851e-4766be8c1588" />

0x4007 commented:
> However, in pool and allowance we need to serve N users simultaneously which means we can hit free RPC rate limits very fast, we need something more robust in my opinion. I think a dynamic RPC is not the proper solution here and permit2-allowance.ubq.fi.

Our rpc handler aggregates from all available rpcs on chain list. It should scale much better than using any one particular rpc from alchemy etc

rndquu reviewed (changes requested):
Transaction URL is broken on successful mint, pls fix:
<img width="1512" alt="Screenshot 2025-02-07 at 17 54 56" src="https://github.com/user-attachments/assets/3919e6b7-fed9-4b02-ad70-2c219e4986c9" />

zugdev commented:
now using rpc-handler here and in permit2-allowance.ubq.fi, this should be the last round of review :)

zugdev commented:
> 1. Fix [failing](https://github.com/ubiquity/uusd.ubq.fi/actions/runs/13255279357/job/37000983458) knip CI

Fixed

> 2. Update README to use `bun` instead of `yarn`

Done

> 3. Transaction URL leads to localhost instead of etherscan:

This is only in localhost (anvil), we map it based on chain:
```
export const explorersUrl: { [key: string]: string } = {
  1: "https://etherscan.io",
  31337: "http://127.0.0.1:8545",
};
```

rndquu reviewed (approved):
Works fine, I've moved a couple of fixes to https://github.com/ubiquity/uusd.ubq.fi/issues/17

EresDev reviewed (changes requested):
- checkbox and label should be on the same line. This is chrome. 
![image](https://github.com/user-attachments/assets/01238a01-a455-409c-8702-fe3d0c6322e7)

- UUSD never showed up in my wallet on mint. I tried this on anvil. On mainnet ethereum, I wasn't able to get through the "Dollar price too low" error because dollar price was less than 1. But I think it should have given me UUSD on anvil. 

- We should remove err part "localhost: check error.stack at console". 
![image](https://github.com/user-attachments/assets/9cd5e2f8-ca5b-4811-bedd-e6e0b77c9e8c)

Some less important suggesions: Can be skipped or turned into separate tasks if other team members agree. 

- let's reuse favicons from https://uad.ubq.fi/ look at its html

- UI for Advanced settings for slippage needs improvement. How about having a settings icons somewhere in mint page (as it is on pancakeswap) and set these settings in the modal? Not having model gives the entire UI a jump on click. Plus, the button itself looks like textbox but is actually a button, so its effect is probably unexpected. By default, they should be filled with default values and not be empty.

- "Force Collateral Only" isn't the term a little unclear? How about two radio buttons or a dropdown with first one selected by default:
[-] Pay with Collateral (LUSD xx amount) and Governance (UBQ  xx amount)
[  ] Pay with Collateral (LUSD xx amount) only
With this maybe we can also get rid of entire "Results" section.

rndquu commented:
> UUSD never showed up in my wallet on mint. I tried this on anvil. On mainnet ethereum, I wasn't able to get through the "Dollar price too low" error because dollar price was less than 1. But I think it should have given me UUSD on anvil.

Follow [README](https://github.com/ubiquity/uusd.ubq.fi/blob/a0a4db6ec98c1211382c61fffd1f695e10c7c58f/README.md) instructions to get past the "Dollar price too low" error

rndquu commented on src/pages/mint.ts:
```diff
@@ -0,0 +1,517 @@
+          isForceCollateralOnlyChecked,
+        });
+
+        const tx = await signerDiamondContract.mintDollar(
```
Setting max slippages was intended behavior to ease minting/redeeming UUSD

EresDev commented:
> Follow [README](https://github.com/ubiquity/uusd.ubq.fi/blob/a0a4db6ec98c1211382c61fffd1f695e10c7c58f/README.md) instructions to get past the "Dollar price too low" error

It was a different problem on different network. 

On local anvil, there was no error because I followed the readme. But UUSD never showed up in my wallet. 

On ethereum mainnet, I had this error "Dollar price too low" , which I think can't be changed.

zugdev commented:
> * checkbox and label should be on the same line. This is chrome.

Will fix

>   ![image](https://private-user-images.githubusercontent.com/11886219/417023506-01238a01-a455-409c-8702-fe3d0c6322e7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA1ODI2ODMsIm5iZiI6MTc0MDU4MjM4MywicGF0aCI6Ii8xMTg4NjIxOS80MTcwMjM1MDYtMDEyMzhhMDEtYTQ1NS00MDljLTg3MDItZmUzZDBjNjMyMmU3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI2VDE1MDYyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY1NjFkMzg4MzYwNTE0ZTVlZjNhMGI5YjkyOWViOGMyN2ZjOWRkMGU4NWEwY2U3MWVhZWVjMzlkMWEzODgwMTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.6I8Tv-yZvm_T2biyFyCbHGxPaA3j5GKU5XZs-w9azZ8)
> * UUSD never showed up in my wallet on mint. I tried this on anvil. On mainnet ethereum, I wasn't able to get through the "Dollar price too low" error because dollar price was less than 1. But I think it should have given me UUSD on anvil.

What wallet are you using? Did you add the custom token in that chain? Perhaps you can try: https://ethereum.stackexchange.com/a/164538

> * We should remove err part "localhost: check error.stack at console".

This is only shown in localhost, my idea is to help further contributors to debug later. Lmk if it's bad.

>   ![image](https://private-user-images.githubusercontent.com/11886219/417025451-9cd5e2f8-ca5b-4811-bedd-e6e0b77c9e8c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA1ODI2ODMsIm5iZiI6MTc0MDU4MjM4MywicGF0aCI6Ii8xMTg4NjIxOS80MTcwMjU0NTEtOWNkNWUyZjgtY2E1Yi00ODExLWJlZGQtZTZlMGI3N2M5ZThjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI2VDE1MDYyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM5MzA5MDRiMjFjYzQ4N2JlZDljNzNiYjhjNjliZTIxMGE5ZmVhMTY1OWFhMTZhOGNiMDA5MDk4YjFjNmRlMTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.tspBB_bOcByMDP1G8_PnOIq97kjJ1Xt18di_YBBTAfY)
> * let's reuse favicons from https://uad.ubq.fi/ look at its html

I see Ubiquity icons is that what you mean?

> 
> Some less important suggesions: Can be skipped or turned into separate tasks if other team members agree.
> 
> * UI for Advanced settings for slippage needs improvement. How about having a settings icons somewhere in mint page (as it is on pancakeswap) and set these settings in the modal? Not having model gives the entire UI a jump on click. Plus, the button itself looks like textbox but is actually a button, so its effect is probably unexpected. By default, they should be filled with default values and not be empty.
> * "Force Collateral Only" isn't the term a little unclear? How about two radio buttons or a dropdown with first one selected by default:
>   [-] Pay with Collateral (LUSD xx amount) and Governance (UBQ  xx amount)
>   [  ] Pay with Collateral (LUSD xx amount) only
>   With this maybe we can also get rid of entire "Results" section.

I like your approach! Current UI is inspired by Frax, which was the one in spec, that's why input, fees, result sections are well separated. On slippage selection I think we could improve it, but making this UI super non-technical might not be the highest priority task. The [Ubiquity Ramp](https://github.com/ubiquity/uusd.ubq.fi/issues/4) will abstract the mint and redeem process entirely, so I think we can append this tasks into #17 

Frax UI, for reference:
<img width="499" alt="image" src="https://github.com/user-attachments/assets/26e6e005-9d5e-4b4a-a60a-7bf8e5492895" />

EresDev commented:
> > * UUSD never showed up in my wallet on mint. I tried this on anvil. On mainnet ethereum, I wasn't able to get through the "Dollar price too low" error because dollar price was less than 1. But I think it should have given me UUSD on anvil.
> 
> What wallet are you using? Did you add the custom token in that chain? Perhaps you can try: https://ethereum.stackexchange.com/a/164538

I have this wallet imported in to metamask and I am trying mint/redeem using this wallet. 
`BENEFICIARY_WALLET=0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266`

Do I need a custom token? It is anvil fork. Shouldn't I receive [UUSD](https://etherscan.io/token/0xb6919Ef2ee4aFC163BC954C5678e2BB570c2D103) on anvil fork of eth mainnet. Do I need to check some other token? Let me know if I am missing anything here.
The redeem worked, and I received UBQ and LUSD. But mint doesn't give me UUSD on anvil eth fork. 

> 
> > * We should remove err part "localhost: check error.stack at console".
> 
> This is only shown in localhost, my idea is to help further contributors to debug later. Lmk if it's bad.

ok then you can keep it. 

> > * let's reuse favicons from https://uad.ubq.fi/ look at its html
> 
> I see Ubiquity icons is that what you mean?

Yes, unless we have some specific favicon

Overall, the code is well written. I tried to break is in many ways and it worked very nicely.

zugdev commented:
Thanks for the compliment!

I feel like you need to add UUSD as a custom token in Metamask. Since you are able to redeem and get UBQ and LUSD, it's most likely that you had UUSD to begin with - else you wouldn't be able to redeem.

Will fix and request your final review, thanks for the attention to details.

EresDev commented:
> > > * UUSD never showed up in my wallet on mint. I tried this on anvil. On mainnet ethereum, I wasn't able to get through the "Dollar price too low" error because dollar price was less than 1. But I think it should have given me UUSD on anvil.
> > 
> > What wallet are you using? Did you add the custom token in that chain? Perhaps you can try: https://ethereum.stackexchange.com/a/164538
> 
> I have this wallet imported in to metamask and I am trying mint using this wallet. 
> `BENEFICIARY_WALLET=0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266`
> 
> Do I need a custom token? It is anvil fork. Shouldn't I receive [UUSD](https://etherscan.io/token/0xb6919Ef2ee4aFC163BC954C5678e2BB570c2D103) on anvil fork of eth mainnet. Do I need to check some other token? Let me know if I am missing anything here.
> The redeem worked, and I received UBQ and LUSD. But mint doesn't give me UUSD on anvil eth fork. 

I just retried. Mint is working with anvil. Sometimes Anvil shows it a little late for me.

zugdev commented:
@EresDev I forked the way `work.ubq.fi` goes about icons, which is probably the latest by 0x4007 and therefore the "most correct" approach. I didn't find `uad.ubq.fi` repo, as it's probably deprecated, it's probably not the best idea to depend on that domain to render the icons. 

1. I fixed the alignment of the checkbox:

<img width="583" alt="image" src="https://github.com/user-attachments/assets/a2311736-8881-4f33-91b4-cf96d2897d67" />

2. Added icons and some meta info:

<img width="219" alt="image" src="https://github.com/user-attachments/assets/9d640567-f99e-4098-b700-c87b3b92559e" />

I think you should open a proposal for your other ideas, that way you can properly be credited and rewarded.
